roboteam_ai/src/control/PositionController.cpp

```diff
@@ -0,0 +1,214 @@
+        velPID.setPID(3.0, 1.0, 0.2);
+    else
+        velPID.setPID(3.0, 0.5, 1.5);
+
+    PIDHasInitialized = true;
+    posVelAngle.vel = error;
+    return pidController(robot, posVelAngle, false);
+}
+
+PosVelAngle PositionController::force(RobotPtr robot, Vector2 &targetPos) {
+    double forceRadius;
+    if ((targetPos - robot->pos).length() < 0.1) {
+        return {robot->pos,{0.0,0.0},0.0};
+    }
+    else if ((targetPos - robot->pos).length() < Constants::MIN_DISTANCE_FOR_FORCE()) {
+        forceRadius = Constants::ROBOT_RADIUS_MAX()*3.0;
+        initializePID(3.0, 1.0, 0.2);
+    }
+    else {
+        forceRadius = Constants::ROBOT_RADIUS_MAX()*6.0;
+        initializePID(3.0, 0.5, 1.5);
+    }
+
+    PosVelAngle target;
+    roboteam_msgs::World world = World::get_world();
+    Vector2 force = (targetPos - robot->pos);
+    force = (force.length() > 3.0) ?
+            force.stretchToLength(3.0) : force;
+
+    for (auto bot : world.us) {
+        force = force
+                + ControlUtils::calculateForce((Vector2) robot->pos - bot.pos, 1, forceRadius);
+    }
+
+    for (auto bot : world.them) {
+        force = force
+                + ControlUtils::calculateForce((Vector2) robot->pos - bot.pos, 1, forceRadius);
+    }
+
+    force = (force.length() > 3.0) ?
+            force.stretchToLength(3.0) : force;
+
+    target.vel = force;
+    return pidController(robot, target);
+}
+
+PosVelAngle PositionController::numTree(RobotPtr robot, Vector2 &targetPos) {
+    PosVelAngle target = numTreeController.goToPos(robot, targetPos);
+    if (target.empty())
+        return force(robot, targetPos);
+    else
+        return pidController(robot, target);
+}
+
+void PositionController::setAvoidBall(bool _avoidBall) {
+    // Add a function to avoid the ball for all goToPos's
+    numTreeController.setAvoidBall(_avoidBall);
+}
+
+void PositionController::setCanGoOutsideField(bool _canGoOutsideField) {
+    // Add a function to make sure the robot does not go out of the field for all goToPos's
+    numTreeController.setCanGoOutsideField(_canGoOutsideField);
+}
+
+PosVelAngle PositionController::pidController(const RobotPtr &robot, PosVelAngle target, bool checkInterface) {
+    PosVelAngle pidCommand;
+    if (! PIDHasInitialized)
+        initializePID();
+    if (checkInterface)
+        checkInterfacePID();
+
+    Vector2 pidP = Vector2();
+    Vector2 pidV = Vector2();
+
+    if (target.pos != Vector2() && ! (posPID.getP() == 0.0 && posPID.getI() == 0.0 && posPID.getD() == 0.0)) {
+        pidP = posPID.controlPIR(target.pos - robot->pos, robot->vel);
+    }
+    if (target.vel != Vector2() && ! (velPID.getP() == 0.0 && velPID.getI() == 0.0 && velPID.getD() == 0.0)) {
+        pidV = velPID.controlPIR(target.vel, robot->vel);
+    }
+
+    pidCommand.pos = target.pos;
+    pidCommand.vel = (pidP + pidV).length() < Constants::MAX_VEL() ?
+                     (pidP + pidV) : (pidP + pidV).stretchToLength(Constants::MAX_VEL());
```

Comment:
You made a one-liner in two lines to not make an if statement why u like dis?

Reply:
it is prettiererer